### PR TITLE
Move PCAP import/export into a plugin

### PR DESCRIPTION
**Pull request by dominiklohmann**
###  :notebook_with_decorative_cover: Description

This (rather large) commit creates two new plugin types, namely `reader_plugin` and `writer_plugin`, and creates a PCAP plugin that implements both. Additionally, it fixes unit tests for plugins, which were unable to actually use the plugin up until now.

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

File-by-file. Ideally this is peer-reviewed in a call.

### CODE REVIEW

File: .github/workflows/static-binary.yaml [hunk 1]
```diff
@@ -48,7 +48,7 @@ jobs:
         fetch-depth: 0
 
     - name: Install Nix
-      uses: cachix/install-nix-action@v12
+      uses: cachix/install-nix-action@v13
```
Comment (dominiklohmann) [inline .github/workflows/static-binary.yaml:51]: Move into separate PR if merging this one takes a while.

File: libvast/src/format/reader_factory.cpp [hunk 2]
```diff
@@ -63,9 +60,30 @@ void factory_traits<format::reader>::initialize() {
...
+        [name = std::string{plugin->name()}](const caf::settings& options)
+          -> caf::expected<std::unique_ptr<format::reader>> {
+          for (const auto& plugin : plugins::get()) {
+            if (plugin->name() != name)
+              continue;
+            const auto* reader = plugin.as<reader_plugin>();
+            VAST_ASSERT(reader);
+            auto in = detail::make_input_stream(options);
+            if (!in)
+              return in.error();
```
Comment (dominiklohmann) [inline libvast/src/format/reader_factory.cpp:76]: Move these into every reader plugins' `make_reader` (and do the same for writer plugins and `make_writer`).

File: plugins/pcap/integration/reference/example/step_00.ref [hunk 3]
```diff
@@ -0,0 +1 @@
+"0.1.0-0"
```
Comment (dominiklohmann) [inline plugins/pcap/integration/reference/example/step_00.ref:1]: Delete file

File: plugins/pcap/integration/reference/example/step_01.ref [hunk 4]
```diff
@@ -0,0 +1 @@
+{"answer":42}
```
Comment (dominiklohmann) [inline plugins/pcap/integration/reference/example/step_01.ref:1]: Delete file

File: plugins/pcap/integration/tests.yaml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      - command: -N import zeek
+        input: data/pcap/zeek/http.log.gz
+      - command: -N import zeek
+        input: data/pcap/zeek/packet_filter.log.gz
+      - command: -N import zeek
+        input: data/pcap/zeek/ssh.log.gz
+      - command: -N pivot --format=zeek zeek.conn '#type ~ /suricata.*/ && dest_ip == 72.247.178.18'
+        transformation: awk '!/^#(open|close)/'
+      - command: -N pivot zeek.conn '#type ~ /zeek.http/ && method == "GET"'
+        transformation: awk '!/^#(open|close)/'
```
Comment (dominiklohmann) [inline plugins/pcap/integration/tests.yaml:61]: Duplicate into regular integration test suite; symlink data directory.

File: plugins/pcap/main.cpp [hunk 6]
```diff
@@ -0,0 +1,796 @@
...
+  size_t flush_interval_ = 0;
+  size_t snaplen_ = 65535;
+  size_t total_packets_ = 0;
+  std::unique_ptr<struct pcap, pcap_close_wrapper> pcap_ = nullptr;
+  std::unique_ptr<struct pcap_dumper, pcap_dump_close_wrapper> dumper_
+    = nullptr;
+  std::string trace_;
+};
+
+/// An example plugin.
```
Comment (dominiklohmann) [inline plugins/pcap/main.cpp:662]: Fix comment

File: plugins/pcap/tests.cpp [hunk 7]
```diff
@@ -3,30 +3,29 @@
 //   | |/ / __ |_\ \  / /          Across
 //   |___/_/ |_/___/ /_/       Space and Time
 //
-// SPDX-FileCopyrightText: (c) 2016 The VAST Contributors
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
 // SPDX-License-Identifier: BSD-3-Clause
 
-#define SUITE format
+#define SUITE pcap_reader
```
Comment (dominiklohmann) [inline plugins/pcap/tests.cpp:9]:
```suggestion
#define SUITE pcap
```

File: plugins/pcap/tests.cpp [hunk 8]
```diff
@@ -3,30 +3,29 @@
...
+#include "vast/format/writer_factory.hpp"
+#include "vast/table_slice.hpp"
+#include "vast/table_slice_builder_factory.hpp"
+#include "vast/table_slice_column.hpp"
 
-#  include <filesystem>
+#include <filesystem>
 
-using namespace vast;
+namespace vast::plugins::pcap_reader {
```
Comment (dominiklohmann) [inline plugins/pcap/tests.cpp:28]:
```suggestion
namespace vast::plugins::pcap {
```

File: plugins/pcap/tests.cpp [hunk 9]
```diff
@@ -57,15 +56,23 @@ std::string_view community_ids[] = {
...
+};
+
 } // namespace
 
 // Technically, we don't need the actor system. However, we do need to
 // initialize the table slice builder factories which happens automatically in
 // the actor system setup. Further, including this fixture gives us access to
 // log files to hunt down bugs faster.
-FIXTURE_SCOPE(pcap_tests, fixtures::deterministic_actor_system)
+FIXTURE_SCOPE(pcap_reader_tests, fixture)
```
Comment (dominiklohmann) [inline plugins/pcap/tests.cpp:73]: Update comment above.

**Review by dominiklohmann — COMMENTED** [review]
Peer-reviewed with @tobim.

File: vast/integration/vast_integration_suite.yaml [hunk 10]
```diff
@@ -283,14 +283,34 @@ tests:
       - command: -N export ascii
         input: queries/multi_addr.txt
 
+  Pivot with PCAP:
+    tags: [node, pivot, pcap]
```
Comment (tobim) [inline vast/integration/vast_integration_suite.yaml:287]: This should just be named `Pivot`, and the `pcap` tag should be removed.

File: vast/integration/vast_integration_suite.yaml [hunk 11]
```diff
@@ -283,14 +283,34 @@ tests:
...
+      - command: -N pivot zeek.conn '#type ~ /zeek.http/ && method == "GET"'
+        transformation: awk '!/^#(open|close)/'
+
   Taxonomy queries:
     tags: [concepts, models]
     fixture: ServerTester
     steps:
       - command: import -b zeek
-        input: data/zeek/pcap-conn.log.gz
+        input: data/pcap/zeek/conn.log.gz
```
Comment (tobim) [inline vast/integration/vast_integration_suite.yaml]: I think we should try improve the structure of the `data` directory. I propose to group all files by format and add a `from-pcap` subdir where appropriate. Then this file would be at `data/zeek/from-pcap/example.conn.log.gz`.

**Review by tobim — CHANGES_REQUESTED** [review]

File: vast/integration/vast_integration_suite.yaml (hunk 11 above)
Comment (dominiklohmann) [reply vast/integration/vast_integration_suite.yaml]: Should we maybe do this in a separate PR?

File: vast/integration/vast_integration_suite.yaml (hunk 10 above)
Comment (dominiklohmann) [reply vast/integration/vast_integration_suite.yaml:287]: Done.

File: vast/integration/vast_integration_suite.yaml (hunk 11 above)
Comment (dominiklohmann) [reply vast/integration/vast_integration_suite.yaml]: I've written a story.

**Review by tobim — APPROVED** [review]
